The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14129294222/job/39585466314) [(plain)](https://github.com/rust-lang/rust/commit/6a4630499d92ac50a3d42f5594eeba0b0816d15f/checks/39585466314/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Attempting with retry: make prepare
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap.py", line 1402, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1382, in main
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1341, in bootstrap
    build.check_vendored_status()
  File "/checkout/src/bootstrap/bootstrap.py", line 1209, in check_vendored_status
    git_merge_commit_email, nightly_branch = self.get_values()
  File "/checkout/src/bootstrap/bootstrap.py", line 1189, in get_values
    with open(file_path, "r") as file:
FileNotFoundError: [Errno 2] No such file or directory: 'src/stage0'
make: *** [Makefile:100: prepare] Error 1
Command failed. Attempt 2/5:
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap.py", line 1402, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1382, in main
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1341, in bootstrap
    build.check_vendored_status()
  File "/checkout/src/bootstrap/bootstrap.py", line 1209, in check_vendored_status
    git_merge_commit_email, nightly_branch = self.get_values()
  File "/checkout/src/bootstrap/bootstrap.py", line 1189, in get_values
    with open(file_path, "r") as file:
FileNotFoundError: [Errno 2] No such file or directory: 'src/stage0'
make: *** [Makefile:100: prepare] Error 1
Command failed. Attempt 3/5:
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap.py", line 1402, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1382, in main
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1341, in bootstrap
    build.check_vendored_status()
  File "/checkout/src/bootstrap/bootstrap.py", line 1209, in check_vendored_status
    git_merge_commit_email, nightly_branch = self.get_values()
  File "/checkout/src/bootstrap/bootstrap.py", line 1189, in get_values
    with open(file_path, "r") as file:
FileNotFoundError: [Errno 2] No such file or directory: 'src/stage0'
make: *** [Makefile:100: prepare] Error 1
Command failed. Attempt 4/5:
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap.py", line 1402, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1382, in main
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1341, in bootstrap
    build.check_vendored_status()
  File "/checkout/src/bootstrap/bootstrap.py", line 1209, in check_vendored_status
    git_merge_commit_email, nightly_branch = self.get_values()
  File "/checkout/src/bootstrap/bootstrap.py", line 1189, in get_values
    with open(file_path, "r") as file:
FileNotFoundError: [Errno 2] No such file or directory: 'src/stage0'
make: *** [Makefile:100: prepare] Error 1
Command failed. Attempt 5/5:
Traceback (most recent call last):
  File "/checkout/src/bootstrap/bootstrap.py", line 1402, in <module>
    main()
  File "/checkout/src/bootstrap/bootstrap.py", line 1382, in main
    bootstrap(args)
  File "/checkout/src/bootstrap/bootstrap.py", line 1341, in bootstrap
    build.check_vendored_status()
  File "/checkout/src/bootstrap/bootstrap.py", line 1209, in check_vendored_status
    git_merge_commit_email, nightly_branch = self.get_values()
  File "/checkout/src/bootstrap/bootstrap.py", line 1189, in get_values
    with open(file_path, "r") as file:
FileNotFoundError: [Errno 2] No such file or directory: 'src/stage0'
make: *** [Makefile:100: prepare] Error 1
The command has failed after 5 attempts.
  local time: Fri Mar 28 12:53:34 UTC 2025
  network time: Fri, 28 Mar 2025 12:53:34 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/src/bootstrap/bootstrap.py
+++ b/src/bootstrap/bootstrap.py
@@ -1162,6 +1162,30 @@ class RustBuild(object):
         config = self.get_toml("build")
         return config or default_build_triple(self.verbose)
 
+    def is_git_repository(self, repo_path):
+        return os.path.isdir(os.path.join(repo_path, ".git"))
+
+    def get_latest_commit(self):
+        repo_path = self.rust_root
+        author_email = self.stage0_data.get("git_merge_commit_email")
+        if not self.is_git_repository(repo_path):
+            return "<commit>"
+        cmd = [
+            "git",
+            "-C",
+            repo_path,
+            "rev-list",
+            "--author",
+            author_email,
+            "-n1",
+            "HEAD",
+        ]
+        try:
+            commit = subprocess.check_output(cmd, universal_newlines=True).strip()
+            return commit or "<commit>"
+        except subprocess.CalledProcessError:
+            return "<commit>"
+
     def check_vendored_status(self):
         """Check that vendoring is configured properly"""
         # keep this consistent with the equivalent check in bootstrap:
@@ -1174,7 +1198,8 @@ class RustBuild(object):
                 eprint("      use vendored sources by default.")
 
         cargo_dir = os.path.join(self.rust_root, ".cargo")
-        url = "https://ci-artifacts.rust-lang.org/rustc-builds/<commit>/rustc-nightly-src.tar.xz"
+        commit = self.get_latest_commit()
+        url = f"https://ci-artifacts.rust-lang.org/rustc-builds/{commit}/rustc-nightly-src.tar.xz"
         if self.use_vendored_sources:
             vendor_dir = os.path.join(self.rust_root, "vendor")
             if not os.path.exists(vendor_dir):

